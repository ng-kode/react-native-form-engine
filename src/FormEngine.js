import React from 'react';
import PropTypes from 'prop-types';
import produce from 'immer';
import {View} from 'react-native';
import lodash from 'lodash';
import validate from 'validate.js';
import defaultTemplates from './templates';

class FormEngine extends React.Component {
  state = {
    errors: {},
    touched: {},
  };

  render() {
    const {
      fields,
      value: currentState,
      verticalSpacing: marginBottom,
      templates,
    } = this.props;
    const {errors, touched} = this.state;
    const makeSpacing = idx =>
      idx === fields.length - 1 ? {} : {marginBottom};

    return (
      <>
        {fields.map((field, idx) => {
          this.throwInvalidField(field);
          const {template, path, customize} = field;
          const Component = templates[template];
          const fromEngine = {
            value: lodash.get(currentState, path),
            onChange: value => this.handleChange(path, value),
            onBlur: () => this.handleBlur(path),
            touched: touched[path],
            errorText: (errors[path] || [])[0],
          };

          return (
            <View key={path} style={makeSpacing(idx)}>
              <Component fromEngine={fromEngine} customize={customize} />
            </View>
          );
        })}
      </>
    );
  }

  throwInvalidField = field => {
    const {template, path} = field;

    if (!path) {
      throw Error(`path is missing for field ${JSON.stringify(field)}`);
    }

    if (!template) {
      throw Error(`template is missing for field ${JSON.stringify(field)}`);
    }
  };

  handleChange = (path, value) => {
    const {value: currentState, onChange} = this.props;
    const nextState = produce(currentState, draftState =>
      lodash.set(draftState, path, value),
    );
    onChange(nextState);
    this.handleValidate(nextState);
  };

  handleValidate = nextState => {
    const {constraints} = this.props;
    const errors = validate(nextState, constraints) || {};
    this.setState({errors});
  };

  handleBlur = path => {
    const touched = produce(this.state.touched, draftTouched => {
      draftTouched[path] = true;
    });
    this.setState({touched});
  };
}

FormEngine.propTypes = {
  value: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  fields: PropTypes.array.isRequired,
  constraints: PropTypes.object,
  verticalSpacing: PropTypes.number,
  templates: PropTypes.object,
};

FormEngine.defaultProps = {
  constraints: {},
  verticalSpacing: 16,
  templates: defaultTemplates,
};

export default FormEngine;
