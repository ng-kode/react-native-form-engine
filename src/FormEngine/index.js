import React from 'react';
import PropTypes from 'prop-types';
import produce from 'immer';
import {View} from 'react-native';
import lodash from 'lodash';
import validate from 'validate.js';
import defaultTemplates from './templates';

class FormEngine extends React.Component {
  state = {
    errors: {},
    touched: {},
    dirty: {},
    isFormDirty: false,
  };

  getVisibleFields(state) {
    const {fields} = this.props;
    return fields.filter(
      ({showOnlyWhen}) => validate(state, showOnlyWhen) === undefined,
    );
  }

  getHiddenFields(state) {
    const {fields} = this.props;
    return fields.filter(
      ({showOnlyWhen}) => validate(state, showOnlyWhen) !== undefined,
    );
  }

  render() {
    const {children, value: currentState} = this.props;
    const visibleFields = this.getVisibleFields(currentState);

    return (
      <>
        {visibleFields.map((field, idx) => {
          this.throwInvalidField(field);
          return this.renderField(field, idx === visibleFields.length - 1);
        })}

        {typeof children === 'function'
          ? this.renderChildrenFunction()
          : children}
      </>
    );
  }

  throwInvalidField = field => {
    const {template} = field;

    if (!template) {
      throw Error(`template is missing for field ${JSON.stringify(field)}`);
    }
  };

  renderField = (field, isLast) => {
    const {template, path, customize: customizeRaw} = field;
    const {
      value: currentState,
      templates,
      verticalSpacing: marginBottom,
    } = this.props;
    const {errors, touched, dirty} = this.state;

    const Component = templates[template];
    const fromEngine = {
      value: lodash.get(currentState, path),
      onChange: value => this.handleChange(path, value),
      onBlur: () => this.handleBlur(path),
      touched: touched[path],
      dirty: dirty[path],
      errorText: (errors[path] || [])[0],
    };

    const customize = Object.keys(customizeRaw).reduce((acc, attr) => {
      const rawValue = customizeRaw[attr];
      if (rawValue.validation !== undefined) {
        const evaluatedValue =
          validate(currentState, rawValue.validation) === undefined;
        return {...acc, [attr]: evaluatedValue};
      } else {
        return {...acc, [attr]: rawValue};
      }
    }, {});

    return (
      <React.Fragment key={path || customize.title}>
        <View style={!isLast && {marginBottom}}>
          <Component fromEngine={fromEngine} customize={customize} />
        </View>
      </React.Fragment>
    );
  };

  renderChildrenFunction = () => {
    const {errors, isFormDirty} = this.state;
    const isFormValid = Object.keys(errors).length === 0;
    return this.props.children({isFormValid, isFormDirty});
  };

  handleChange = (path, value) => {
    this.setState({
      isFormDirty: true,
      dirty: {...this.state.dirty, [path]: true},
    });
    const {value: currentState, onChange} = this.props;

    let nextState = produce(currentState, draftState => {
      lodash.set(draftState, path, value);
    });

    const fieldsToHide = this.getHiddenFields(nextState);

    nextState = produce(nextState, draftState => {
      fieldsToHide.forEach(field => {
        lodash.unset(draftState, field.path);
      });
    });

    onChange(nextState);
    this.handleValidate(nextState);
  };

  handleValidate = nextState => {
    const {validation} = this.props;
    const errors = validate(nextState, validation) || {};
    this.setState({errors});
  };

  handleBlur = path => {
    const touched = produce(this.state.touched, draftTouched => {
      draftTouched[path] = true;
    });
    this.setState({touched});
  };
}

FormEngine.propTypes = {
  value: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  fields: PropTypes.array.isRequired,
  validation: PropTypes.object,
  verticalSpacing: PropTypes.number,
  templates: PropTypes.object,
};

FormEngine.defaultProps = {
  validation: {},
  verticalSpacing: 16,
  templates: defaultTemplates,
};

export default FormEngine;
