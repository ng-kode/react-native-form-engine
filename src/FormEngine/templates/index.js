import TextField from './TextField';
import PickerField from './PickerField';
import SwitchField from './SwitchField';
import Text from './Text';

const templates = {
  TextField,
  PickerField,
  SwitchField,
  Text,
};

export default templates;
