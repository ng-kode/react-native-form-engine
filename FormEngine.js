import React from 'react';
import PropTypes from 'prop-types';
import produce from 'immer';
import lodash from 'lodash';
import validate from 'validate.js';
import templates from './templates';

const constraints = {
  email: {email: {message: '^INVALID_EMAIL'}},
  preferredOS: {exclusion: ['windows']},
};

class FormEngine extends React.Component {
  state = {
    errors: {},
    touched: {},
  };

  render() {
    const {fields, value: currentState} = this.props;
    const {errors, touched} = this.state;

    return (
      <>
        {fields.map(field => {
          this.throwInvalidField(field);
          const {template, path, options} = field;
          const Component = templates[template];
          const fieldValue = lodash.get(currentState, path);
          const errorText = (errors[path] || [])[0];

          return (
            <Component
              key={path}
              value={fieldValue}
              onChange={value => this.handleChange(path, value)}
              onBlur={() => this.handleBlur(path)}
              errorText={errorText}
              touched={touched[path]}
              options={options}
            />
          );
        })}
      </>
    );
  }

  throwInvalidField = field => {
    const {template, path, options} = field;

    if (!path) {
      throw Error(`path is missing for field ${JSON.stringify(field)}`);
    }

    if (!template) {
      throw Error(`template is missing for field ${JSON.stringify(field)}`);
    }
  };

  handleChange = (path, value) => {
    const {value: currentState, onChange} = this.props;
    const nextState = produce(currentState, draftState =>
      lodash.set(draftState, path, value),
    );
    onChange(nextState);
    this.handleValidate(nextState);
  };

  handleValidate = nextState => {
    const errors = validate(nextState, constraints) || {};
    this.setState({errors});
  };

  handleBlur = path => {
    const touched = produce(this.state.touched, draftTouched => {
      draftTouched[path] = true;
    });
    this.setState({touched});

    const {value: currentState} = this.props;
    this.handleValidate(currentState);
  };
}

FormEngine.propTypes = {
  value: PropTypes.object.isRequired,
  onChange: PropTypes.func.isRequired,
  fields: PropTypes.array.isRequired,
};

export default FormEngine;
