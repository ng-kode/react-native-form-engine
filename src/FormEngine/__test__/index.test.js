// import 'react-native';
import React from 'react';
import FormEngine from '../index';

import TestRenderer from 'react-test-renderer';

describe('FormEngine', () => {
  test('should render null if fields = []', () => {
    const fields = [];
    const json = TestRenderer.create(
      <FormEngine value={{}} onChange={() => {}} fields={fields} />,
    ).toJSON();
    expect(json).toBe(null);
  });

  test('should render custom fields', () => {
    const fields = [
      {path: 'username', template: 'TextField'},
      {path: 'email', template: 'TextField'},
    ];
    const json = TestRenderer.create(
      <FormEngine value={{}} onChange={() => {}} fields={fields} />,
    ).toJSON();
    expect(json.length).toBe(2);
    expect(json).toMatchSnapshot();
  });

  test('should render fields with values referenced from `value` props', () => {
    const fields = [
      {path: 'username', template: 'TextField'},
      {path: 'email', template: 'TextField'},
    ];
    const value = {username: 'super_123', email: 'thisisme@example.com'};
    const json = TestRenderer.create(
      <FormEngine value={value} onChange={() => {}} fields={fields} />,
    ).toJSON();
    expect(json).toMatchSnapshot();
  });

  test('should provide each custom field with props `fromEngine` and `customize`', () => {
    const fields = [
      {path: 'username', template: 'TextField'},
      {path: 'email', template: 'TextField'},
    ];
    const Fields = TestRenderer.create(
      <FormEngine value={{}} onChange={() => {}} fields={fields} />,
    ).root.findAllByProps({testID: 'form-engine-field'});
    expect(Fields.every(f => f.props.fromEngine !== undefined)).toBe(true);
    expect(Fields.every(f => f.props.customize !== undefined)).toBe(true);
  });

  test('should trigger onChange when template props onChange is called', () => {
    const fields = [
      {path: 'username', template: 'TextField'},
      {path: 'email', template: 'TextField'},
    ];
    const onChange = jest.fn();
    const firstTextInput = TestRenderer.create(
      <FormEngine value={{}} onChange={onChange} fields={fields} />,
    ).root.findAllByProps({testID: 'form-engine-field'})[0];

    firstTextInput.props.fromEngine.onChange('user123');
    expect(onChange).toBeCalledTimes(1);
    expect(onChange).toBeCalledWith({username: 'user123'});
  });
});
