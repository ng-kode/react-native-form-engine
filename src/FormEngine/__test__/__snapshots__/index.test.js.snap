// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormEngine should render custom fields 1`] = `
Array [
  <View
    style={
      Object {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginBottom": 4,
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "lightgrey",
            "borderRadius": 4,
            "borderWidth": 1,
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </View>,
  <View
    style={false}
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginBottom": 4,
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "lightgrey",
            "borderRadius": 4,
            "borderWidth": 1,
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
      />
    </View>
  </View>,
]
`;

exports[`FormEngine should render fields with values referenced from \`value\` props 1`] = `
Array [
  <View
    style={
      Object {
        "marginBottom": 16,
      }
    }
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginBottom": 4,
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "lightgrey",
            "borderRadius": 4,
            "borderWidth": 1,
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
        value="super_123"
      />
    </View>
  </View>,
  <View
    style={false}
  >
    <View
      style={Object {}}
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "marginBottom": 4,
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        editable={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "borderColor": "lightgrey",
            "borderRadius": 4,
            "borderWidth": 1,
            "padding": 8,
          }
        }
        underlineColorAndroid="transparent"
        value="thisisme@example.com"
      />
    </View>
  </View>,
]
`;
